python/pfs/utils/optimizeBlackDots.py

```diff
@@ -0,0 +1,535 @@
+            y coordinate around which to rotate
+
+        Returns
+        ----------
+        xd_new, yd_new: `float`, `float`
+            x and y coordinate of the new position
+
+        Notes
+        ---------
+        This function allows the user to move the dots in x and y direction,
+        and rotate/scale around custom origin
+
+        Calls `rotate`
+        """
+        dots = self.dots
+
+        xd_original = dots.loc[i-1]['x']
+        yd_original = dots.loc[i-1]['y']
+
+        xd_offset = xd_original - x_scale_rot
+        yd_offset = yd_original - y_scale_rot
+
+        r_distance_from_center = np.sqrt(xd_offset**2 + yd_offset**2)
+
+        xc_rotated, yc_rotated = self.rotate((xd_offset*(1+scale*r_distance_from_center/100),
+                                              yd_offset*(1+scale*r_distance_from_center/100)), (0, 0), rot)
+
+        xc_rotated = xc_rotated+x_scale_rot
+        yc_rotated = yc_rotated+y_scale_rot
+
+        xd_new = xc_rotated + xd_mod
+        yd_new = yc_rotated + yd_mod
+        return xd_new, yd_new
+
+    def quality_measure_single_spot(self, list_of_prepared_observations_and_predictions_single_fiber,
+                                    xd, yd):
+        """Calculate a penalty given x and y dot positions
+
+        Parameters
+        ----------
+        list_of_prepared_observations_and_predictions: `list`
+            list contaning observations for only one fiber
+        xd: `float`
+            x position of the dot
+        yd: `float`
+            y position of the dot
+
+        Returns
+        ----------
+        quality_measure: `float`
+            total penalty for this position of the dot
+
+        Notes
+        ---------
+        Calculates `penalty`. Penalty consist of 2 factors:
+            1. Points should have been seen given the dot position,
+            but they are not seen
+            2. Point which are seen, but they should not have been
+            seen, given the dot position
+
+        Calls `check`
+        Gets called by `total_penalty_for_single_dot`
+        """
+        list_of_quality_measure_single_run = []
+        for i in range(len(list_of_prepared_observations_and_predictions_single_fiber)):
+            prepared_observations_and_predictions =\
+                list_of_prepared_observations_and_predictions_single_fiber[i]
+
+            actual_position_x_observed,\
+                actual_position_y_observed,\
+                predicted_position_x_not_observed,\
+                predicted_position_y_not_observed = prepared_observations_and_predictions
+
+            # the ones which were seen - add penalty if you should not have seen them
+            quality_measure_observed = []
+            for j in range(len(actual_position_x_observed)):
+                quality_measure_observed.append(self.check(actual_position_x_observed[j],
+                                                           actual_position_y_observed[j], xd, yd, 0.75))
```

Comment:
The `0.75` is the default value in the method. Don't repeat it here. If you need to use a value here, it should be a configuration parameter.